### PR TITLE
I1541 refactor and rejig optimization

### PR DESCRIPTION
**Pull request by BhargavaKrishnaSreepathi**
This PR adds in the functions identified during the executive course.
This is also the first step in refactoring a significant portion of Optimization
Multiprocessing is implemented in this PR, and the computational time for a case study went down from 5000 secs to 1300 secs (about a quarter of time) (wooohoooooo).

The way to test this PR.
1. Take any of the case studies, either heating or cooling. 
2. Follow the procedure of datahelper, radiation, demand, potentials, followed by optimization_main.py
3. Notable changes to config are the inclusion of 
district-heating-network = true
district-cooling-network = false
Just make sure, one of these is true

### CODE REVIEW

**Comment by BhargavaKrishnaSreepathi** [comment]
local tests run on my computer

File: cea/optimization/optimization_main.py [hunk 1]
```diff
@@ -10,10 +10,12 @@
 import cea.inputlocator
 from cea.optimization.prices import Prices as Prices
 import cea.optimization.distribution.network_opt_main as network_opt
-import cea.optimization.master.master_main as master
+import cea.optimization.master.master_main as new_master
```
Comment (daren-thomas) [inline cea/optimization/optimization_main.py]: why? why `new_master`? what's "new" about it? in 6 months time, will it still be "new"? do we need to import ... "as"? only do this if relying heavily (multiple times) on the contents of the imported module. also, what is wrong with the name `master_main`? at least that is the same as the module...

File: cea/optimization/optimization_main.py [hunk 2]
```diff
@@ -53,7 +55,7 @@ def moo_optimization(locator, weather_file, gv, config):
     # pre-process information regarding resources and technologies (they are treated before the optimization)
     # optimize best systems for every individual building (they will compete against a district distribution solution)
     print "PRE-PROCESSING"
-    extra_costs, extra_CO2, extra_primary_energy, solarFeat = preproccessing(locator, total_demand, building_names,
+    extra_costs, extra_CO2, extra_primary_energy, solar_features = preproccessing(locator, total_demand, building_names,
```
Comment (daren-thomas) [inline cea/optimization/optimization_main.py]: nice! i like this!

File: cea/optimization/optimization_main.py [hunk 3]
```diff
@@ -63,8 +65,8 @@ def moo_optimization(locator, weather_file, gv, config):
 
     # optimize conversion systems
     print "CONVERSION AND STORAGE OPTIMIZATION"
-    master.evolutionary_algo_main(locator, building_names, extra_costs, extra_CO2, extra_primary_energy, solarFeat,
-                                  network_features, gv, config, prices, lca)
+    new_master.master_main(locator, building_names, extra_costs, extra_CO2, extra_primary_energy, solar_features,
```
Comment (daren-thomas) [inline cea/optimization/optimization_main.py]: oh. i see. you have a function called `master_main` inside the module `master_main`. and... and what does that function do? i think the function might be able to do with a better name, what do you think?

File: cea/optimization/master/master_main.py [hunk 4]
```diff
@@ -65,8 +65,8 @@ def objective_function_wrapper(args):
     Wrap arguments because multiprocessing only accepts one argument for the function"""
     return objective_function(*args)
 
-def master_main(locator, building_names, extra_costs, extra_CO2, extra_primary_energy, solar_features,
-                network_features, gv, config, prices, lca):
+def non_dominated_sorting_genetic_algorithm(locator, building_names, extra_costs, extra_CO2, extra_primary_energy, solar_features,
```
Comment (daren-thomas) [inline cea/optimization/master/master_main.py]: now that's fancy naming! awesome, man!

File: cea/optimization/optimization_main.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 import cea.inputlocator
 from cea.optimization.prices import Prices as Prices
 import cea.optimization.distribution.network_opt_main as network_opt
-import cea.optimization.master.master_main as new_master
+import cea.optimization.master.master_main as master
```
Comment (daren-thomas) [inline cea/optimization/optimization_main.py]: you're still renaming the module name...

**Comment by BhargavaKrishnaSreepathi** [comment]
@daren-thomas: fixed the bug you encountered. I had to add the corresponding parameters in the cli interface.

**Comment by daren-thomas** [comment]
@BhargavaKrishnaSreepathi what is the status of this PR? How do I test it?

File: cea/technologies/chiller_absorption.py [hunk 6]
```diff
@@ -250,10 +220,11 @@ def calc_Cinv(qcold_W, locator, ACH_type, config):
...
                 InvC = Inv_a + Inv_b * (Q_nom_each_chiller) ** Inv_c + (Inv_d + Inv_e * Q_nom_each_chiller) * log(Q_nom_each_chiller)
                 Capex_a1 = InvC * (Inv_IR) * (1 + Inv_IR) ** Inv_LT / ((1 + Inv_IR) ** Inv_LT - 1)
-                Capex_a = Capex_a + Capex_a1
-                Opex_fixed = Opex_fixed + Capex_a1 * Inv_OM
+                Capex_a_ACH_USD = Capex_a_ACH_USD + Capex_a1
+                Opex_fixed_ACH_USD = Opex_fixed_ACH_USD + Capex_a1 * Inv_OM
+                Capex_ACH_USD = Capex_ACH_USD + InvC
 
-    return Capex_a, Opex_fixed
+    return Capex_a_ACH_USD, Opex_fixed_ACH_USD, Capex_ACH_USD
```
Comment (daren-thomas) [inline cea/technologies/chiller_absorption.py]: This changes the return type of the function from a 2-tuple to a 3-tuple. That changes the functions signature. When doing this, make sure to update all consumers of this function! You can find them with ALT+F7

All other consumers of this function (see below) are now broken...

image

**Comment by BhargavaKrishnaSreepathi** [comment]
Currently, no where. May be we should do the dependencies somewhere in the documentation

**Comment by daren-thomas** [comment]
Are you ready to merge yet?

**Comment by BhargavaKrishnaSreepathi** [comment]
hey @daren-thomas, this is ready to be merged. Please test it based on the procedure I mentioned at the top of the PR. Local tests run on my computer

**Comment by daren-thomas** [comment]
For documentation purposes... the following graph can be created with graphviz as in here: https://gist.github.com/daren-thomas/8fbc90506f906925adb9aa05dec286e1

image

the file `docs/configuration-file-details.rst` contains an example of using graphviz in our documentation. Adapting to this graph, would be:

```
.. digraph:: "cea-optimization-dependencies"

    rankdir="TD";
    node [shape=box];
   
    "network-layout" -> "thermal-network-matrix" -> "optimization" -> "multi-criteria-analysis";
    "decentralized" -> "optimization";
```

**Comment by daren-thomas** [comment]
Hi @BhargavaKrishnaSreepathi I'm trying to create the documentation and am getting this error:

```
In [1]: import cea.optimization.individual_evaluation.individual_cooling_main
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-2d2b8faba952> in <module>()
----> 1 import cea.optimization.individual_evaluation.individual_cooling_main

C:\Users\darthoma\Documents\GitHub\CityEnergyAnalyst\cea\optimization\individual_evaluation\individual_cooling_main.py in <module>()
     24 from cea.technologies.thermal_network.thermal_network_matrix import calculate_ground_temperature
     25 from cea.constants import WH_TO_J
---> 26 from cea.optimization.constants import EL_TO_CO2, EL_TO_OIL_EQ, SIZING_MARGIN, PUMP_ETA, DELTA_U, \
     27     ACH_T_IN_FROM_CHP, ACH_TYPE_DOUBLE, T_TANK_FULLY_CHARGED_K, T_TANK_FULLY_DISCHARGED_K, PEAK_LOAD_RATIO, \
     28     NG_CC_TO_CO2_STD, NG_CC_TO_OIL_STD

ImportError: cannot import name EL_TO_CO2
```

What changed?

**Comment by daren-thomas** [comment]
Optimization is failing. Here are the steps I performed:

- del c:\reference-case-open\baseline
- cea extract-reference-case
- cea demand --scenario c:\reference-case-open\baseline --weather Zug --region CH
- cea network-layout --network-type DH
- cea thermal-network-matrix --network-type DH
- cea decentralized
- cea photovoltaic
- cea solar-collector --type-scpanel FP
- cea solar-collector --type-scpanel ET
- cea photovoltaic-thermal --type-scpanel FP
- cea sewage-potential
- cea lake-potential
- cea optimization

```
Traceback (most recent call last):
  File "C:\Users\darthoma\AppData\Local\conda\conda\envs\cea\Scripts\cea-script.py", line 11, in <module>
    load_entry_point('cityenergyanalyst', 'console_scripts', 'cea')()
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\interfaces\cli\cli.py", line 65, in main
    script_module.main(config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\optimization\optimization_main.py", line 118, in main
    moo_optimization(locator=locator, weather_file=weather_file, gv=gv, config=config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\optimization\optimization_main.py", line 66, in moo_optimization
    network_features, gv, config, prices, lca)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\optimization\master\master_main.py", line 157, in non_dominated_sorting_genetic_algorithm
    repeat(prices, len(invalid_ind)), repeat(lca, len(invalid_ind))))
  File "c:\users\darthoma\appdata\local\conda\conda\envs\cea\lib\multiprocessing\pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "c:\users\darthoma\appdata\local\conda\conda\envs\cea\lib\multiprocessing\pool.py", line 567, in get
    raise self._value
AttributeError: 'SolarFeatures' object has no attribute 'A_PV_m2'
```

**Comment by BhargavaKrishnaSreepathi** [comment]
Hey @daren-thomas, I followed the exact procedure you did and did not get the error. I suspect, in the config file you have the flag for district heating and district cooling as false.

**Comment by daren-thomas** [comment]
Hi @BhargavaKrishnaSreepathi I am now getting the following error with `multi-criteria-analysis` - after re-running the `optimization` script:

```
λ cea multi-criteria-analysis
City Energy Analyst version 2.9.0
Running `cea multi-criteria-analysis` with the following parameters:
- general:scenario = c:\reference-case-open\baseline
- general:region = CH
- multi-criteria:generations = 2
- multi-criteria:network-type = DC
- multi-criteria:economicsustainability = 0.8
- multi-criteria:environmentalsustainability = 0.1
- multi-criteria:socialsustainability = 0.1
- multi-criteria:annualizedcosts = 0.8
- multi-criteria:capextotal = 0.1
- multi-criteria:opex = 0.1
- multi-criteria:emissions = 0.5
- multi-criteria:primaryenergy = 0.5
- multi-criteria:renewableshare = 1.0
Running multicriteria with scenario = c:\reference-case-open\baseline
Running multicriteria for generation = 2
Traceback (most recent call last):
  File "C:\Users\darthoma\AppData\Local\conda\conda\envs\cea\Scripts\cea-script.py", line 11, in <module>
    load_entry_point('cityenergyanalyst', 'console_scripts', 'cea')()
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\interfaces\cli\cli.py", line 65, in main
    script_module.main(config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\analysis\multicriteria\main.py", line 541, in main
    multi_criteria_main(locator, config)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\analysis\multicriteria\main.py", line 44, in multi_criteria_main
    data_address = locating_individuals_in_generation_script(generation, locator)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\analysis\multicriteria\optimization_post_processing\locating_individuals_in_generation_script.py", line 60, in locating_individuals_in_generation_script
    data_generation = preprocessing_generations_data(locator, generation)
  File "c:\users\darthoma\documents\github\cityenergyanalyst\cea\analysis\multicriteria\optimization_post_processing\locating_individuals_in_generation_script.py", line 141, in preprocessing_generations_data
    data['population_fitness']]  # convert to millions
KeyError: 'population_fitness'
```

**Pull request merged**